cpp/core/utils/Registry.h

```diff
@@ -0,0 +1,52 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <mutex>
+#include <string>
+#include <unordered_map>
+
+#include "utils/Exception.h"
+
+namespace gluten {
+template <typename T>
+class Registry {
+ public:
+  void registerObj(const std::string& kind, T t) {
+    std::lock_guard<std::mutex> l(mutex_);
+    GLUTEN_CHECK(map_.find(kind) == map_.end(), "Already registered for " + kind);
+    map_[kind] = std::move(t);
+  }
+
+  T& get(const std::string& kind) {
+    std::lock_guard<std::mutex> l(mutex_);
```

Comment:
I note these are code refactors. Does read operation deserve to be optimized with `std::shared_lock` to allow concurrency?

Reply:
No need to do that so far. We can optimize once it's really required.